### ISSUE TITLE
Enhancement: Observing Dependency Properties

### ISSUE THREAD

**Issue by MovGP0** [runceel/ReactiveProperty#39]
It should be possible to convert an `DependencyProperty` to an Observable (and Reactive Property) in WPF und UWP projects. 

The first step would be to get notifications when the property changes: 

```csharp
    public static class DependencyObjectExtensions
    {
        public static IObservable<EventArgs> Observe<T>(this T component, DependencyProperty dependencyProperty)
            where T:DependencyObject
        {
            return Observable.Create<EventArgs>(observer =>
            {
                EventHandler update = (sender, args) => observer.OnNext(args);
                var property = DependencyPropertyDescriptor.FromProperty(dependencyProperty, typeof(T));
                property.AddValueChanged(component, update);
                return Disposable.Create(() => property.RemoveValueChanged(component, update));
            });
        }
    }
```

I guess a similar approach could be used that additionally uses the `SetValue(DependencyProperty, (object)T)` and `(T)GetValue(DependencyProperty)` methods to wrap the `DependencyProperty` with an `ReactiveProperty`. 

```csharp
        public static ReactiveProperty<TValue> ToReactiveProperty<TDependencyObject, TValue>(this TDependencyObject component, DependencyProperty dependencyProperty)
            where TDependencyObject:DependencyObject
        {
            var initialValue = (TValue) component.GetValue(dependencyProperty);
            var property = new ReactiveProperty<TValue>(initialValue /* TODO: needs IScheduler and ReactivePropertyMode */);

            // subscribe to changes of the reactive property
            var changes = component.Observe(dependencyProperty);

            // TODO: needs to get disposed when ReactiveProperty gets disposed
            var subscription = changes.Subscribe(args => property.Value = (TValue)component.GetValue(dependencyProperty));
            
            return property;
        }
``

**Comment by runceel** [comment]
I think that this is good.
But, I concern TODO in ToReactiveProperty.

**Comment by MovGP0** [comment]
I think that there is the need for an `internal void Add(IDisposable disposable)` method in the `ReactiveProperty` class, in order to ensure that the subscription gets disposed, when the `ReactiveProperty` gets disposed.

**Comment by runceel** [comment]
I trying implement.
But, It couldn't implement in UWP.(nothing DependencyPropertyDescriptor)

I will release this methods WPF only.

**Comment by runceel** [comment]
By the way, When this method use?

I want to know use case.

**Comment by MovGP0** [comment]
The issue is when I want to observe changes of the value of a data-bound `DependentProperty` in the view and update the `ViewModel`. 

Ie. when the `TextProperty` changes and the `ViewModel` has to recalculate some measurements of the control.

**Comment by MovGP0** [comment]
**Note** 
There is an alternative way of doing this in WPF, which should also work in UWP: override the property with `DependentProperty.OverrideMetadata(...)` in order to be able to register a `PropertyChangedCallback` handler in the `DependencyPropertyMetadata`. 

However, I find it quite messy, since it introduces code duplication of code from framework base classes. using `DependencyPropertyDescriptor` to observe changes globally is a nicer solution. 

The advantage is that I can update the `ViewModel` of the `DependencyObject` directly via the `PropertyChangedCallback`.

**Comment by runceel** [comment]
What is difference with data binding + converter?

**Comment by MovGP0** [comment]
With the `DependencyPropertyDescriptor` I can directly access the property changes from the object instance and don't need to know all the specifics of that property. 

With `OverrideMetadata` the implementation gets replaced, so I need to know how the dependency property is implemented, which usually requires decompiling. Then I need to duplicate that implementation and add my handler. But since this is static, I can only get my DependencyObject by typecasting the `object sender` parameter, and then access the `DataContext` to get the `ViewModel`.

**Comment by MovGP0** [comment]
maybe you can elaborate how the data binding with the converter should work. For a reuseable control I need dependency properties (DP) in the code behind (`MyControl.xaml.cs`) file, as well as data binding to the ViewModel (VM) in the control template (ie. in `Generic.xaml`). The DP of the user control gets bound to some other data context of the parent control (ie. `MyPage.xaml`). 

When the DP of the parent control changes, the DP of the user control changes too (because of the data binding) and the VM of the user control needs to react to those changes. 

My current architecture looks like the following: 
![reactivepropertyarchitecture](https://cloud.githubusercontent.com/assets/83339/20734167/e80fb422-b697-11e6-8a6d-d3f26cec810b.png)

**Comment by runceel** [comment]
Thank you for provide usecase.

I released v3.4.0.
I added ObservableDependencyProperty, ToReadOnlyReactiveProperty, ToReactiveProperty extension method to DependencyObject.

https://www.nuget.org/packages/ReactiveProperty/3.4.0
